src/Catel.MVVM/Catel.MVVM.Shared/Collections/FastObservableCollection.cs

```diff
@@ -427,31 +428,26 @@ protected void NotifyChanges()
         {
             Action action = () =>
             {
-                // Create event args
-                NotifyCollectionChangedEventArgs eventArgs = null;
-                if (_suspensionContext != null && _suspensionContext.Mode == SuspensionMode.Adding)
+                List<NotifyCollectionChangedEventArgs> eventArgsList = new List<NotifyCollectionChangedEventArgs>();
+
+                if (_suspensionContext != null)
                 {
                     if (_suspensionContext.NewItems.Count != 0)
                     {
-                        eventArgs = CreateEventArgs(NotifyCollectionChangedAction.Add, _suspensionContext.NewItems, _suspensionContext.NewItemIndices);
+                        eventArgsList.Add(CreateEventArgs(NotifyCollectionChangedAction.Add, _suspensionContext.NewItems, _suspensionContext.NewItemIndices));
                     }
-                }
-                else if (_suspensionContext != null && _suspensionContext.Mode == SuspensionMode.Removing)
-                {
+
                     if (_suspensionContext.OldItems.Count != 0)
                     {
-                        eventArgs = CreateEventArgs(NotifyCollectionChangedAction.Remove, _suspensionContext.OldItems, _suspensionContext.OldItemIndices);
+                        eventArgsList.Add(CreateEventArgs(NotifyCollectionChangedAction.Remove, _suspensionContext.OldItems, _suspensionContext.OldItemIndices));
                     }
                 }
                 else
                 {
-                    //Debug.Assert(_suspensionContext != null && _suspensionContext.Mode == SuspensionMode.None, "Wrong/unknown suspension mode!");
-
-                    eventArgs = CreateEventArgs(NotifyCollectionChangedAction.Reset);
+                    eventArgsList.Add(CreateEventArgs(NotifyCollectionChangedAction.Reset));
                 }
 
-                // Fire events
-                if (eventArgs != null)
+                foreach (var eventArgs in eventArgsList)
```

Comment:
I don't think we have to raise Count and Item[] for every change. I think this should be the logic:

1. If Reset => just clear everything and do a full reset
2. If only added / removed => single event args (combine data)
3. At the end, raise Count + Item[] just once